File: game.py
from flask import Flask, jsonify, request
from flask_cors import CORS
import game_service
from db_initializer import initialize_database

app = Flask(__name__)
CORS(app)

current_game_id = None
current_player_id = None
refresh_counter = 0


@app.route('/start', methods=['POST'])
def start_game():
    global current_game_id, current_player_id, refresh_counter
    try:
        data = request.json or {}
        pilot_name = data.get('name', 'Captain Anonymous') #gets pilot's name.

        initialize_database() #refreshes database
        game_service.initialize_fuel_prices() #gets fuel price

        current_game_id = game_service.create_game() #creates game entry

        current_player_id = game_service.create_player(current_game_id, pilot_name) # creates player

        refresh_counter = 0 #at the start of the game sets the counter at 0.

        return jsonify({
            "status": "success",
            "message": f"Welcome aboard, Captain {pilot_name}!",
            "player_id": current_player_id
        })
    except Exception as e:
        print(f"Error in /start: {e}")
        return jsonify({"status": "error", "message": str(e)}), 500


@app.route('/status', methods=['GET'])
def get_status():

    if not current_game_id or not current_player_id:
        return jsonify({"error": "No active game session found. Please start a new game."}), 400

    try:
        turns_left = game_service.get_turns_left(current_game_id)

        player = game_service.show_player(current_player_id, turns_left) # player structure: [name, money, fuel, current_airport_ident, city_name]. can be found in show_player function.

        ident = player[3]   # Extract location details ident = 4th column's value in the show_player data.
        city_name = player[4] # city_name is the fifth column's value.

        coords = game_service.get_airport_coordinates(ident)
        fuel_price = game_service.get_fuel_price(ident)

        contracts = game_service.get_contracts(current_player_id)

        return jsonify({
            "pilot": player[0],
            "money": player[1],
            "fuel": player[2],
            "turns": turns_left,
            "location": ident,
            "city": city_name,
            "fuel_price": fuel_price,
            "coords": {"lat": coords[0], "lng": coords[1]},
            "contracts": contracts,
            "max_fuel": game_service.MAX_FUEL,
            "goal_money": game_service.TARGET_MONEY
        })
    except Exception as e:
        print(f"Error in /status: {e}")
        return jsonify({"error": "Failed to retrieve game status"}), 500

@app.route('/fly', methods=['POST'])
def fly():
    try:
        data = request.json
        contract = data.get('contract')

        if not contract:
            return jsonify({"success": False, "message": "No contract selected."}), 400


        player = game_service.show_player(current_player_id, 0)
        if player[2] < contract['fuel_needed']:  # player[2] is the fuel level. check if the current fuel level is less than the fuel req.
            return jsonify({"success": False, "message": "Not enough fuel for this flight!"})
        msg = game_service.update_player_after_contract(current_game_id, current_player_id, contract)

        game_service.reduce_turn(current_game_id) #deduct a turn

        return jsonify({"success": True, "message": msg})
    except Exception as e:
        print(f"Error in /fly: {e}")
        return jsonify({"success": False, "message": "Flight system failure."}), 500




@app.route('/refuel', methods=['POST'])
def refuel():

    try:
        data = request.json
        try:
            amount = int(data.get('amount'))
        except (TypeError, ValueError):
            amount = 0

        if amount <= 0:
            return jsonify({"success": False, "message": "Please enter a positive fuel amount."})

        msg = game_service.refuel_player(current_game_id, current_player_id, amount)

        success = msg.startswith(
            "Refueled")  # if the return message after refuel_player function starts with Refueled.
        return jsonify({"success": success, "message": msg})

    except ValueError:
        return jsonify({"success": False, "message": "Invalid number format for fuel."})
    except Exception as e:
        print(f"Error in /refuel: {e}")
        return jsonify({"success": False, "message": "Refueling system offline."}), 500

if __name__ == '__main__':
    app.run(debug=True, port=5000)

File: game.js
const map = L.map('map', { zoomControl: false }).setView([60.2, 24.9], 5);
L.tileLayer('https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
    attribution: '© OpenStreetMap contributors'
}).addTo(map); // map initialization.
const markersGroup = L.layerGroup().addTo(map);

let playerName = "";


function nextStep() {
    document.getElementById('welcome-screen').classList.add('hidden');
    document.getElementById('name-screen').classList.remove('hidden');
}

async function finishSetup() {
    playerName = document.getElementById('player-name-input').value;
    if (!playerName) return alert("Captain, we need your name!");

    try {
        const response = await fetch('http://127.0.0.1:5000/start', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ name: playerName })
        });
        document.getElementById('name-screen').classList.add('hidden');
        const briefing = document.getElementById('briefing-screen');
        briefing.querySelector('h1').innerText = `Welcome, Captain ${playerName}!`;
        briefing.querySelector('.goal-text').innerHTML =
            `Budget : Fuel - 800 units and Money - 1000€ <br><br>Mission: Reach 3000€ before turns run out.`;
        briefing.classList.remove('hidden');
    } catch (e) {
        alert("Server connection failed.");
    }
}

function enterGame() {
    document.body.classList.remove('pre-game');
    document.getElementById('setup-overlay').classList.add('hidden');
    updateStatus();
}


async function updateStatus() {
    try {
        const response = await fetch('http://127.0.0.1:5000/status');
        const data = await response.json();

        // 1. Check Win/Loss Conditions
        if (data.money >= data.goal_money) {
            alert(`VICTORY! You reached the goal of ${data.money}€ and won the game!`);
            location.reload();
            return;
        }
        if (data.turns <= 0) {
            alert(`GAME OVER! You ran out of turns. Final money: ${data.money}€. Try again!!`);
            location.reload();
            return;
        }

        // 2. Update HUD Text & Bars
        document.getElementById('money-val').innerText = data.money;
        document.getElementById('fuel-val').innerText = data.fuel;
        document.getElementById('turn-val').innerText = data.turns;

        document.getElementById('fuel-bar').style.width = (data.fuel / data.max_fuel * 100) + "%";
        document.getElementById('money-bar').style.width = (data.money / data.goal_money * 100) + "%";
        document.getElementById('turn-bar').style.width = (data.turns / 15 * 100) + "%";

        // 3. Update Map
        markersGroup.clearLayers();
        const lat = parseFloat(data.coords.lat);
        const lng = parseFloat(data.coords.lng);

        L.circle([lat, lng], {
            color: 'red',
            fillColor: 'red',
            fillOpacity: 1,
            radius: 30000 // Size in meters
        }).addTo(markersGroup).bindPopup(`<b>Location:</b> ${data.city}<br><br><b>Fuel Price:</b> ${data.fuel_price}`);

        map.flyTo([lat, lng], 4);

        renderContracts(data.contracts); // Updates Contracts using renderContracts which is written below.
        console.log(data.contracts);

    } catch (e) {
        console.error("Status update failed:", e);
    }
}

function renderContracts(contracts) {
    const list = document.getElementById('contract-list');
    if (!list) return;
    list.innerHTML = "";

    contracts.forEach(c => {
        const lat = parseFloat(c.lat);
        const lng = parseFloat(c.lng);

        if (!isNaN(lat) && !isNaN(lng)) {
            const contractData = encodeURIComponent(JSON.stringify(c));

            // Generate content for the Map Popup
            const popupContent = `
                <div style="color: #333; min-width: 140px;">
                    <b style="font-size: 1.1em;">${c.destination_name}</b><br>
                    <hr style="margin: 5px 0; border: 0; border-top: 1px solid #eee;">
                    Distance: ${c.distance} km<br>
                    Fuel Req: ${c.fuel_needed} L<br>
                    Reward: ${c.reward} €<br>
                    <button class="btn-action" onclick="fly('${contractData}')" 
                        style="background: #2ecc71; color: black; font-weight: bold; margin-top: 8px;">
                        FLY HERE
                    </button>
                </div>
            `;

            L.circle([lat, lng], {
                color: '#006400',
                fillColor: '#006400',
                fillOpacity: 1,
                radius: 30000
            }).addTo(markersGroup).bindPopup(popupContent);
        }

    });
}

async function fly(encodedContract) {
    const contract = JSON.parse(decodeURIComponent(encodedContract));
    try {
        const response = await fetch('http://127.0.0.1:5000/fly', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ contract: contract })
        });
        const result = await response.json();
        if (result.success) {
            updateStatus();
        } else {
            alert(result.message);
        }
    } catch (e) {
        alert("Flight system failure!");
    }
}


uiRefreshStatus = () => updateStatus();

uiRefreshContracts = async () => {
    const response = await fetch('http://127.0.0.1:5000/refresh', { method: 'POST' });
    const result = await response.json();
    alert(result.message);
    updateStatus();
};

uiRefuel = async () => {
    const amount = prompt("How many units of fuel do you want to buy?");
    if (!amount || isNaN(amount)) return alert('invalid input');
    const response = await fetch('http://127.0.0.1:5000/refuel', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ amount: parseInt(amount) })
    });
    const result = await response.json();
    alert(result.message);
    updateStatus(); // I would love to ensure that the contracts didn't change when I refueled.
};